### DISCUSSION TITLE
"Facilitation of trade and barter"

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#66] [Nomad Network]
I saw this line on the Feature Roadmap of Nomad Network: "Facilitation of trade and barter" - https://unsigned.io/nomadnet/

Just wanted to say that one of the things that got me thinking about Reticulum was to use it as a network specifically for sending [Grin slatepack](https://docs.grin.mw/wiki/transactions/slatepack/) messages around.

Grin is sort of unique in that you either need both parties to be online at the same time, or you need to send these slatepacks back and forth a couple of times. 

A slatepack message looks like this in text form

```
BEGINSLATEPACK. 4H1qx1wHe668tFW yC2gfL8PPd8kSgv
pcXQhyRkHbyKHZg GN75o7uWoT3dkib R2tj1fFGN2FoRLY
GWmtgsneoXf7N4D uVWuyZSamPhfF1u AHRaYWvhF7jQvKx
wNJAc7qmVm9JVcm NJLEw4k5BU7jY6S eb. ENDSLATEPACK.
```

Pretty cool way to transact, I think. Of course, different currencies and systems will have different solutions for doing this sort of thing, though.

Anyway, just sharing some thoughts I had for how I might utilize Reticulum in the future. Basically combining Reticulum with a crypto wallet in the same app. 

Anyone else have ideas about "facilitation of trade and barter"?

**Comment by 4c3e** [comment]
Wow very interesting! So users can manually send those text messages back and forth via LoRa/ packet radio etc, without the explicit use of a blockchain?
I'm not too active in the digital currency space, but I do know that something like Chaumian ecash could also potentially work over reticulum.

**Reply by trosel** [reply]
> without the explicit use of a blockchain?

The transaction still finalizes on the blockchain of course! 

> something like Chaumian ecash could also potentially work over reticulum.

Oh yes, this project is chaumian mints for bitcoin: https://github.com/fedimint/minimint

**Comment by trosel** [comment]
Now I'm realizing that @markqvist probably just meant the ability to post "for sale" ads and things on Reticulum. Not necessarily the transaction part haha

**Reply by markqvist** [reply]
No, no I did actually mean the whole transaction shebang ;) I just haven't railed much about it anywhere, since I didn't want to attract much attention about it. Sometimes it seems you just have to whisper "cryptocurrency" in a dark alley somewhere, and the whole [Parade from Paprika](https://www.youtube.com/watch?v=FAnSKN9s7eY) amasses.

In my utopian naivety I've actually been trying to find a solution that would also let the finalization layer exist on (even low-bandwidth) Reticulum networks (maybe I should just start calling them *rets* instead of "reticulum networks*).

Grin is insanely interesting. When I read your post I was like, oh let me just check this thing out quickly and reply on the discussion, and then went straight down the rabbit hole of their docs. While I still don't have a complete grasp on the elliptic curve math they use (especially for "collapsing" the transactions into combined kernels), I can definitely say that the system as a whole is completely brilliant. Very, very cool stuff. The slatepack concept really got me thinking.

I've been dazing around with this on my mind for a couple of days now, and haven't been able to get it out of my head. For one thing, it would definitely be possible to run the whole Grin transaction building process over Reticulum, even in a very elegant and native way, since the slatepack "sender" and "receivers" can be natively represented as Destinations on a reticulum network (both are really just Curve25519 points).

But then it also hit me that we could actually build a very neat transaction+finality system directly on top of Reticulum, by using a combination of concepts from [Scrit](https://github.com/scritcash/scrit-whitepaper/raw/master/scrit-whitepaper.pdf) and Grin, and a few concepts. You could actually do this entirely without a blockchain per-se, having the entire network state represented only as a database of unspent outputs. Such a thing would be *very* efficient and easy to sync.

Just for fun, I tried to design a small, high-level, concept-system around this idea, and I think it actually could work. Or I might have missed a bunch of critical details, and it is all just happy pipe dreams. But I think it is worth exploring some more, and I will definitely do so when I have the time. If anyone is interested, I can try to do a write up of the concept and ideas so far. I would be interested to hear your thoughts.

The system I have in mind basically aims at making it possible, maybe even *easy*, to bootstrap economic activity for a community, while ensuring a high degree of privacy, security and distribution of control. Take note of the word *bootstrap* here. This needs to work from zero, with very little setup cost (both social and capital), and still provide reliability and security.

I think this is very important to be able to set up a transaction/value storage system *from scratch* where and when you need it. This is not usually possible with existing blockchains, since you need *huge* capital expenditure to have any semblance of security, i.e., you can fork any chain you like, but you are then completely vulnerable, since someone from that chain can just shift 0.0001% of their hash-power towards your chain and completely control it.

**Reply by trosel** [reply]
> Grin is insanely interesting. 

I agree! I'm no good at math, but from what I can tell, it is all a very elegant and simple system. Simplicity is often too overlooked in crypto projects, in my opinion. 

But in addition to the cool slatepack system, you have the constant emission of 1 grin per second, forever. That is a mind-bending emission strategy. And it is one that really resonates with my life philosophy too. Fair, predictable, reliable, etc.

> For one thing, it would definitely be possible to run the whole Grin transaction building process over Reticulum, even in a very elegant and native way, since the slatepack "sender" and "receivers" can be natively represented as Destinations on a reticulum network (both are really just Curve25519 points).

Yes, that is exactly what I had in mind! But I must confess that I'm not a strong backend developer, so I wasn't exactly sure how to combine the codebases up until now haha. I suppose that one would add the LXMF SDK to a Grin wallet app.

**Scrit** seems really interesting. But is the idea proven to work yet? How far did they get with [this implementation](https://github.com/scritcash/scrit)?

> I think this is very important to be able to set up a transaction/value storage system from scratch where and when you need it. This is not usually possible with existing blockchains, since you need huge capital expenditure to have any semblance of security

Of course the idea with blockchain is that everyone has access to the same global network. So you would never need to fork anything just to have access. But can you explain a little more about why you think a community may want to fork and create a new currency just for themselves?

**Reply by markqvist** [reply]
> I agree! I'm no good at math, but from what I can tell, it is all a very elegant and simple system. Simplicity is often too overlooked in crypto projects, in my opinion.
 
I agree. There's no shortage of overly complex crypto projects, and seeing a really elegant and simple system like Grin is actually a bit of a relief. 

> But in addition to the cool slatepack system, you have the constant emission of 1 grin per second, forever. That is a mind-bending emission strategy. And it is one that really resonates with my life philosophy too. Fair, predictable, reliable, etc.
 
Couldn't agree more.

> Yes, that is exactly what I had in mind! But I must confess that I'm not a strong backend developer, so I wasn't exactly sure how to combine the codebases up until now haha. I suppose that one would add the LXMF SDK to a Grin wallet app.

Don't take my word on this as gospel, since I haven't read their full implementation. This is just what I *believe* from my initial understanding. But from what I understand, you could build the entire slatepack message just from the EC point math already in Reticulum, and then send it to some node on the Grin network for finalization. You would of course need to hook into a Grin wallet somehow to get the signing keys for actually building the slatepack message, and I don't know exactly how, or even if, there is an API for that, but I can't imagine that it's not possible.

Or you could simply add Reticulum as a transport option to the Grin wallet, and have the wallet itself build the slatepack, only doing it over Reticulum instead of HTTPS or Tor. From what I understand, Grin already has a system in place for supporting multiple transports for building transactions, so maybe this is the best way to go about something like that. When I get more time I am probably going to mess around with it, but for now, my hands are full with all the scheduled work on RNS and the new RNode ecosystem.
 
> **Scrit** seems really interesting. But is the idea proven to work yet? How far did they get with [this implementation](https://github.com/scritcash/scrit)?

I actually have no idea on how far their implementation is, and neither how well proven the system is in practice. My guess is that it is more or less completely unproven, practically. But theoretically it should work very well. I really like their approach.

> Of course the idea with blockchain is that everyone has access to the same global network. So you would never need to fork anything just to have access. But can you explain a little more about why you think a community may want to fork and create a new currency just for themselves?

I didn't word that too well. I think it's almost always a pretty bad solution to fork an existing cryptocurrency or chain. The point I tried to make was really, that doing so is almost always completely useless, but situations exists, where it would be really useful to just spin up a reliable transaction layer.

Having the ability to bootstrap economic activity and fungibility within an arbitrary domain, *from zero* and with (virtually) zero cost is a really great tool, that I think could be useful in situations like:

- When you need the ability to transact and trade on networks that are not connected to the Internet.
- When you want to transact reliably, but completely outside the scrutiny of the outside world, or only with a well-defined group of peers.
- Organisation of peer-to-peer loans, debt and repayment.
- Allocation of resources, responsibilities, rights, services and similar, within organisations, communities or groups.
- Ownership management of any kind of *stuff*.
- When you want complete self-reliance for things that are inherently internal to your community (annoying that you can't re-allocate fuel or water, because the Internet is unavailable, or some node needs to re-sync a 300GB chain over a slow connection).

From a social perspective I also think there is a very important aspect in self-ownership over essential tools and infrastructure. If people of a community know that they are themselves in co-ownership and co-control over essentials like the economic transaction system, it directly makes people feel empowered. This breeds individual responsibility, and also helps create a context of meaning. That is important empowerment to return to a community. It's a small positive feedback loop you can start. The more of those you can start, the stronger a community becomes.

Long-term, having self-sovereign transaction and fungibility layers are an important building block of self-sovereign societies, and a necessary precursor to any kind of communally created order-structure. You put that on an "outsourced" tech layer, and what you are really doing is putting your order-structures into the hands of the owners of that tech layer (exactly as it stands in our world now).

**Reply by trosel** [reply]
> Or you could simply add Reticulum as a transport option to the Grin wallet

This is what I was thinking of. Basically you could create a wallet that behaves more like Venmo or Cashapp if you had a messaging layer. To me, that is the end goal! With simple swipe gestures to manage the phases of signing slatepacks.

> Having the ability to bootstrap economic activity and fungibility within an arbitrary domain, from zero and with (virtually) zero cost is a really great tool

Yes, I totally see that! Tokens on the Lightning network come to mind. The value of the tokens only come from *who* is issuing them, which seems accurate to how value really works. Of course the value from a global blockchain is that the "who" is "the whole world" / "everyone committing hash power" :)

**Reply by yeastplume** [reply]
> Don't take my word on this as gospel, since I haven't read their full implementation. This is just what I _believe_ from my initial understanding. But from what I understand, you could build the entire slatepack message just from the EC point math already in Reticulum, and then send it to some node on the Grin network for finalization. You would of course need to hook into a Grin wallet somehow to get the signing keys for actually building the slatepack message, and I don't know exactly how, or even if, there is an API for that, but I can't imagine that it's not possible.

There is indeed an extensive grin-wallet API docs for which can be found at https://docs.rs/grin_wallet_api/latest/grin_wallet_api/. The project is in Rust (and the wallet lib can be linked directly,) but there's also an option to start up a wallet as a secure listener and communicate with the wallet API via JSON-RPC.

Nice to see the interest in Grin, and some appreciation for its main goals of minimalism, fair-launch and fair emission. And reciprocally, we're always in search of a better network for performing slatepack exchanges. If you have any questions or would just like to talk more, feel free to get in touch!

**Reply by markqvist** [reply]
Ah, thanks for pointing this out. I'm going to build some prototype out of that when time allows it :) Thanks a lot for offering your availability to answer questions, and likewise please feel free to ask as well!

The next while for me will be focused on the remaining missing pieces in the Reticulum ecosystem, but here is to hoping that we might be able to collaborate in some way in the not too distant future. The more I look into Grin, the more I like it. I would love to see the ability to perform slatepack exchanges over Reticulum.

**Reply by trosel** [reply]
posting Grin wallet API tutorials here for future reference: https://github.com/grincc/tutorial

**Reply by trosel** [reply]
> The project is in Rust

Bumping this to say they there is now also a Python implementation https://github.com/grinventions/mimblewimble-py

Perhaps easier to work with since Reticulum is also in Python

**Comment by 4c3e** [comment]
Semi-related: It would be really interesting if we generalize the "re-use" of Curve25519 points used in reticulum destinations. I imagine this would have tons of uses outside of grin, for example it could potentially enable the mapping of RNS destinations to interoperate with [qaul.net destinations](https://github.com/qaul/qaul.net/blob/main/docs/qaul.net-manual/security.md#cryptographic-key). (Or any other routing protocol that uses curve25519 destinations!). I need to read a bit more into it, but we already have the existing coordinates for use in reticulum, so it should be fairly straightforward to reuse them in other cryptographic routing protocols. 

The value proposition of the qaul interoperability, imo, is that qaul.net optimizes for high-bandwidth wifi transports, so it's able to do almost-real-time ad-hoc mesh reorganization and spend a lot more bandwidth on routing announcements. It would act as an emergency transport to increase utility in crowds of mobile peers.

**Reply by trosel** [reply]
I suppose that is a question a bit beyond Reticulum itself. 

For example, do you want to use the same endpoint for multiple services and let the context figure out what you want to do with it? Or do you want to have a unique endpoint for multiple services and have everything mapped in a third party app (something like Linktree) where you keep your addresses?

I tend to favor using real life metaphors and patterns wherever possible. Because of that, i think having multiple keys stored on a keyring that you keep secure is the most user friendly at the moment. Rather than some magical routing and context-determining happening programmatically at least.

**Reply by 4c3e** [reply]
yeah good point. The only situation where it might be interesting would be some sort of messaging app where you could switch on qaulnet as an "emergency mode"  or something, so everyone's contacts are already transferred over. Of course, you could also just pre-share everyone's separate qaulnet keys when you add each other as contacts so maybe not that useful, but a little fun! :)
